Base/Python/slicer/tests/test_slicer_parameter_pack.py

```diff
@@ -0,0 +1,288 @@
+import unittest
+from typing import Annotated
+
+from MRMLCorePython import vtkMRMLModelNode
+
+import slicer
+from slicer.parameterNodeWrapper import *
+from slicer.parameterNodeWrapper.parameterPack import ParameterPackSerializer
+
+
+def newParameterNode():
+    node = slicer.mrmlScene.AddNewNodeByClass("vtkMRMLScriptedModuleNode")
+    return node
+
+
+@parameterPack
+class Point:
+    x: float
+    y: float
+
+
+@parameterPack
+class BoundingBox:
+    topLeft: Point
+    bottomRight: Point
+
+
+class TypedParameterNodeTest(unittest.TestCase):
+    def setUp(self):
+        slicer.mrmlScene.Clear(0)
+
+    def test_parameter_pack(self):
+        @parameterPack
+        class ParameterPack:
+            x: int
+            y: Annotated[int, Default(1)]
+
+        pack = ParameterPack()
+        self.assertEqual(pack.x, 0)
+        self.assertEqual(pack.y, 1)
+
+        pack2 = ParameterPack(2, 3)
+        self.assertEqual(pack2.x, 2)
+        self.assertEqual(pack2.y, 3)
+
+        pack3 = ParameterPack(y=2, x=3)
+        self.assertEqual(pack3.x, 3)
+        self.assertEqual(pack3.y, 2)
+
+        pack4 = ParameterPack(6, y=8)
+        self.assertEqual(pack4.x, 6)
+        self.assertEqual(pack4.y, 8)
+
+        pack5 = ParameterPack(6)
+        self.assertEqual(pack5.x, 6)
+        self.assertEqual(pack5.y, 1)
+
+        pack6 = ParameterPack(y=6)
+        self.assertEqual(pack6.x, 0)
+        self.assertEqual(pack6.y, 6)
+
+        with self.assertRaises(ValueError):
+            pack.x = None
+        with self.assertRaises(ValueError):
+            pack.x = "hi"
+
```

Comment:
Might be good to also ensure you can't dynamically add an attribute to the pack; `pack.z = 123`.

Reply:
Good idea, will add.

Reply:
Upon further consideration, do we want this?

`@dataclasses.dataclass` doesn't do this.

Are you trying to protect from accidents (typos) or something malicious?

Reply:
I don't think it's a good thing to support... dataclass does let you assign new attributes but `dataclasses.asdict`, `dataclasses.update`, etc. and the generated methods like `__str__`, `__eq__`, etc. don't see them. I'd seriously question any code that does this; it probably needs a dictionary, not a dataclass.

The value that `@parameterPack` adds over `@dataclass` is its validation and serialization features; supporting dynamic attributes seems problematic at best. You can still put a `dict[str, T]` in a `parameterPack` so I don't think there's any missing functionality.

Reply:
There is difference between supporting and explicitly disallowing. The problem with explicitly disallowing things in Python is that you can't.

I can override `__setattr__` to disallow dynamic addition, but it would add complexity. If I don't disallow it, I would still say it is not supported and people shouldn't do it. Even if I did override `__setattr__`, someone could still do something like `object.__setattr__(pack, "dynamicAdd", someValue)`.

So the question is, is the added complexity worth it. If your concern is people breaking the intended use, I think it is not worth it because people can do that anyway. If the concern is protecting someone typing `pack.valeu = 1` instead of `pack.value = 1`, then maybe (although that is a problem everywhere else in Python too).

Reply:
I see. Probably not worth adding any extra complexity in `__setattr__`; maybe just let it be something that's implicitly _possible_, but not encouraged or guaranteed.